dali/c_api/c_api.h

```diff
@@ -69,6 +75,44 @@ extern "C" {
       int cpu_prefetch_queue_depth,
       int gpu_prefetch_queue_depth);
 
+  /**
+   * @brief Feed the data to ExternalSource as contiguous memory.
+   *
+   * @param pipe_handle Pointer to pipeline handle
+   * @param name Pointer to a null terminated string with the name of the External Source to be fed
+   * @param device Device of the supplied memory. Only CPU is supported.
+   * @param data_ptr Pointer to contiguous buffer containing all samples
+   * @param data_type Type of the provided data
+   * @param shape Pointer to an array containing shape of all samples concatenated one after
+   *              another. Should contain batch_size * sample_dim elements.
+   * @param sample_dim The dimensionality of the data.
+   * @param layout_str Optional layout provided as a pointer null terminated string.
+   *                   Can be set to NULL.
+   */
+  DLL_PUBLIC void daliSetExternalInput(daliPipelineHandle* pipe_handle, const char* name,
+                                       device_type_t device, const void* data_ptr,
+                                       dali_data_type_t data_type, const int64_t* shape,
+                                       int sample_dim, const char* layout_str);
+
+  /**
+   * @brief Feed the data to ExternalSource as a set of separate buffers.
+   *
+   * @param pipe_handle Pointer to pipeline handle
+   * @param name Pointer to a null terminated string with the name of the External Source to be fed
+   * @param device Device of the supplied memory. Only CPU is supported.
+   * @param data_ptr Pointer to an array containing batch_size pointers to separate Tensors.
+   * @param data_type Type of the provided data
+   * @param shape Pointer to an array containing shape of all samples concatenated one after
+   *              another. Should contain batch_size * sample_dim elements.
+   * @param sample_dim The dimensionality of the data.
+   * @param layout_str Optional layout provided as a pointer null terminated string.
+   *                   Can be set to NULL.
+   */
+  DLL_PUBLIC void daliSetExternalInputTensors(daliPipelineHandle* pipe_handle, const char* name,
+                                              device_type_t device, const void* const* data_ptr,
+                                              dali_data_type_t data_type, const int64_t* shape,
+                                              int64_t sample_dim, const char* layout_str);
```

Comment:
I'm not an expert in C APIs, but I think that together with const char* you should require length of that string, in case there are nulls inside. Or it's common to require only one null always at the end?

Reply:
We can go the way of `snprintf`, but I'm not sure if we really want to pile more length related arguments here. NULL terminated strings are quite common.

Reply:
Fair enough ;)